contracts/core/base/ExtraStorageBased.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {KeyValue} from "./../types/Types.sol";
+import {ExtraDataLib} from "../libraries/ExtraDataLib.sol";
+
+abstract contract ExtraStorageBased {
+    using ExtraDataLib for mapping(bytes32 => bytes);
+
+    event Lens_ExtraDataSet(address indexed addr, uint256 indexed entityId, bytes32 indexed key, bytes value);
+
+    /*
+     * ExtraStorage is organized like this:
+     * address Address => uint256 EntityId => bytes32 Key => bytes ABIEncodedValue
```

Comment:
In my opinion, we should also add `bytes32 entityType` to be ready for multi-entity primitives. Yeah, I know, right now all primitives have just a single entity, but in the future there could be multi-entity ones and it would be nice to have a standard already set that allows setting/querying data taking the entity type into account